### PR TITLE
Build Tensorflow and libtensorflow from source

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Restore Tensorflow source build while leaving binary builds . Based on @yorickvP and @timokau work.with some parts taken from both.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @uri-canva for changes regaring Bazel.

Tested both CUDA and non-CUDA builds of the Python library, both source and binary builds. Didn't test libtensorflow, I assume that if Python library works and `ldd` output is okay libtensorflow is okay too.

### CODE REVIEW

**Comment by abbradar** [comment]
Forgot to mention actual previous PRs: #63208 and #63616.

**Comment by abbradar** [comment]
Darwin still uses binary builds for now. I don't have a Darwin machine so I can't test if source build works and fix it if it doesn't - unless someone wants to take on that role!

**Comment by abbradar** [comment]
(For some reason I cannot request a review from @yorickvP on Github)

**Comment by abbradar** [comment]
Updated binary `libtensorflow` doesn't build for Darwin. Can someone with Darwin knowledge help with it?

**Comment by danieldk** [comment]
> Updated binary `libtensorflow` doesn't build for Darwin. Can someone with Darwin knowledge help with it?

It seems that Google now build on newer macOS versions, I had the same problem with Tensorflow 1.13.1 when building on my own MacBook with Mojave. A newer cctools is needed, see #49371. This PR is probably needed: #61172

**Comment by yorickvP** [comment]
This is great! I'm hoping to get a fully static build at some point, but having the same protobuf on tensorflow as the rest of the code should fix our linker errors! :)

You can't request a review from me because I don't have commit rights, I suppose.

File: pkgs/build-support/build-bazel-package/default.nix
Comment (yorickvP) [inline pkgs/build-support/build-bazel-package/default.nix]: how did you manage to eliminate the nix hacks?

**Review by yorickvP — APPROVED** [review]
whoever built tensorflow upstream must have patched out that saved_model_portable_proto line. it's been in there for several version snow.

**Comment by timokau** [comment]
As I mentioned in https://github.com/NixOS/nixpkgs/pull/63616, there are still issues with the python2 build. Seems to be a general problem with parallelism on python2, only manifests itself here because its such a big project and basically hopeless to build without a lot of parallelism (used 12 cores).

I already added one patch to python2 (https://github.com/NixOS/nixpkgs/pull/64067), but I still saw the issue (sometimes, its transient) after that. I suspect the second part of the fix in the python ticket is also needed. That was never backported by debian, so I had to do it myself (being not very experienced with C). I've pushed those changes to the branch of the other PR. Its not very polished, which is why I didn't push it before.

I don't have time to work on this or give extensive review right now. But I can answer questions if you have any.

File: pkgs/build-support/build-bazel-package/default.nix
Comment (abbradar) [reply pkgs/build-support/build-bazel-package/default.nix]: I didn't actually eliminate them, instead I force-enabled them by default for all Nix Bazel builds :D They are needed in the builder to ensure `NIX_*` environment variables always get passed.

**Comment by abbradar** [comment]
@timokau I saw that one but IIUC this is a general Python problem, not Tensorflow-specific one, so I didn't really look into it. It shouldn't happen during the "big" Tensorflow build because I only build "source" package now and build the wheel separately with our `buildPythonProgram`. I didn't observe the problem on 8-core (16-thread) machine during ~10 builds too.

**Comment by timokau** [comment]
Neat, so you only need bazel for the C part and don't need to unnecessarily rebuild everything  for libtensorflow, py2 tensorflow and py3 tensorflow?

**Comment by abbradar** [comment]
Sadly no, you still need to rebuild everything because their C libraries link to different Pythons and I presume use different APIs. Still, in the end you get something that resembles a usual source Python package with `setup.py` etc. I use this as `src` for `buildPythonPackage` so any Python bugs should be isolated there.

**Comment by timokau** [comment]
Oh, so you mean you just skip the final `bdist_wheel` as it was done previously? I actually changed that on purpose, so that we can unify the binary and source builds. That way we wouldn't have all the duplication between `default.nix` and `bin.nix`.

I don't think the `bdist_wheel` step alone was responsible for the build failures, but who knows.

**Comment by abbradar** [comment]
The primary reason I switched back to the previous behavior was to ensure `fixupPhase` runs on unpacked Tensorflow libraries. This ensures RPATH is set up correctly.

**Comment by yorickvP** [comment]
Will this work with bazel 0.28.0? https://github.com/NixOS/nixpkgs/pull/64633

**Comment by timokau** [comment]
My PR had some issues with bazel 0.27, not sure if @abbradar worked around those. But the bazel team has promised no further backwards incompatibilities for the next 3 releases (then they'll break backwards compatibility once more and release 1.0).

**Comment by abbradar** [comment]
> But the bazel team has promised no further backwards incompatibilities for the next 3 releases (then they'll break backwards compatibility once more and release 1.0).

Nice news! I didn't test 0.28 though yet, wanted to do it today but sadly I won't have time for one more rebuild.

> some issues with bazel 0.27, not sure if @abbradar worked around those

Huh, didn't notice anything. Maybe you remember something specific? I apply patch for 0.27 from upstream and also add a flag to enable backwards compatibility with some behavior.

**Comment by timokau** [comment]
@abbradar IIRC there were some python2 issues, I think caused by a mix of bazel moving to python toolchains and 27d4ce7ce08613ab124d160ff269c48c86eb3d83.

**Comment by timokau** [comment]
That was deterministic though, so if py2-tensorflow builds now with current master and this PR it should be fine.

**Comment by timokau** [comment]
Here's the details of the bazel plans: https://blog.bazel.build/2019/06/06/Bazel-Semantic-Versioning.html

Hopefully it won't be as much of a pain anymore from now on.

**Comment by abbradar** [comment]
@timokau I feel quite stupid right now - I completely forgot to test this with `python2`.

Stay tuned...

**Comment by timokau** [comment]
Yeah, forgot that for a while too until I noticed it suddenly wasn't working anymore. Can't wait until we can drop support for python2.

**Comment by abbradar** [comment]
With https://github.com/NixOS/nixpkgs/pull/64900 and a partial revert of https://github.com/NixOS/nixpkgs/pull/53988 Python 2 version builds fine for me too. I'm rebuilding it the third time just to be sure but it was perfectly reproducible without the .pyc patch.

File: pkgs/development/tools/build-managers/bazel/nix-hacks.patch
Comment (uri-canva) [inline pkgs/development/tools/build-managers/bazel/nix-hacks.patch]: Why is ignoring the markers not necessary anymore? I thought it was necessary because we patch the markers to make them deterministic, but was it only necessary in the past maybe because we deleted the markers instead of patching them? I'm not comfortable with removing code without understand why it's there in the first place, can you find out?

File: pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch
Comment (uri-canva) [inline pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch]: Why was this removed? Or rather, why was this removed, but the shebang line substitution was left in? Both force the built Bazel to hardcode the nix version of python in the built python binary. I'm not sure that was intentional, and I feel like that should be part of `enableNixHacks` instead of part of the regular build, but I don't think we should change this as part of this PR.

File: pkgs/build-support/build-bazel-package/default.nix
Comment (uri-canva) [inline pkgs/build-support/build-bazel-package/default.nix]: Agreed with removing all these as commented in https://github.com/NixOS/nixpkgs/pull/56033#issuecomment-511799691.

**Review by uri-canva — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch
Comment (abbradar) [reply pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch]: The shebang line substitution is for the builder script itself which finds an appropriate Python binary; it's reasonable to pin this one as part of a usual build. On the other hand `NIX_STORE_PYTHON_PATH` was used to pin the Python binary and this is problematic for Python 2 builds.

Comment (abbradar) [reply pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch]: See also: https://github.com/NixOS/nixpkgs/pull/64903

File: pkgs/development/tools/build-managers/bazel/nix-hacks.patch
Comment (abbradar) [reply pkgs/development/tools/build-managers/bazel/nix-hacks.patch]: Indeed, this works now because we just delete the markers instead of patching them. I suppose when I implemented this back in fe153d73ce3add3f0733469c40b1748cd0743e5f and ba9cde1dd55532f673c3fd64bce62290c19b4e32 I either didn't try just removing the markers (?) or somehow it didn't work.

Comment (abbradar) [reply pkgs/development/tools/build-managers/bazel/nix-hacks.patch]: > 2. Build an actual derivation which uses fetched dependencies (skipping checksums needed here because they depend on the build directory).
This might also be a hint, maybe Bazel changed their checksum algorithm since then to not depend on something? Maybe inodes?

Comment (uri-canva) [reply pkgs/development/tools/build-managers/bazel/nix-hacks.patch]: I expect deleting the markers to not work, but patching them correctly should. Sounds good to me if you've tested it and it doesn't cause bazel to download the dependencies again after the initial fetch.

Comment (abbradar) [reply pkgs/development/tools/build-managers/bazel/nix-hacks.patch]: BTW see also https://github.com/NixOS/nixpkgs/pull/64904 , I splitted it to a separate PR.

File: pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch
Comment (uri-canva) [reply pkgs/development/tools/build-managers/bazel/python-stub-path-fix.patch]: I see, so it's less about the nix version or the system version, and more about 2 vs 3. Ok, I have no issues with it then!

**Review by uri-canva — APPROVED** [review]
Bazel changes look good to me.

**Comment by Profpatsch** [comment]
@GrahamcOfBorg build bazel.tests

**Comment by Profpatsch** [comment]
The changes to bazel look indeed good, all tests pass.

**Comment by timokau** [comment]
> I'm rebuilding it the third time just to be sure but it was perfectly reproducible without the .pyc patch.

I'm a bit confused by this. So you had no issues without the python patch in https://github.com/NixOS/nixpkgs/pull/64900? Why add it then?

**Comment by abbradar** [comment]
Nono, the problem was that I somehow messed up I Python 2 and 3, so I thought that I evaded the bug because Python 3 worked. I still feel quite stupid :D Most likely because build times are huge so I had a lot of context switches between my attempts.

**Comment by timokau** [comment]
So you mean that you're trying to verify that it's perfectly reproducible **with** the pyc patch?

For what it's worth for verification I used my [nix-rebuild](https://github.com/timokau/dotfiles/blob/08be2f95a09f0e2a679daedbeb79001d6fbcefe8/scripts/bin/nix-rebuild) script like this:

```
try=0
while nix-rebuild python2.pkgs.tensorflow >logs/$try.log 2>&1; do
    echo "$try done"
    (( try += 1 ))
````

And just let that run overnight.

**Comment by abbradar** [comment]
Yup, I'm actually running something similar right now and want to leave it overnight too.

**Comment by timokau** [comment]
> In the message above I meant that without the patch I couldn't build it successfully even once before

Oh, now I understand. For me it wasn't entirely deterministic, which was annoying. That's why I prematurely concluded https://github.com/NixOS/nixpkgs/pull/64067 had fixed the issue.

**Comment by abbradar** [comment]
Now that other parts are here I'll wait till staging goes into master and merge this (so that we don't break Python 2 + Tensorflow).

**Comment by timokau** [comment]
Why not just retarget this PR to staging and merge now? Then we'll get it automatically on the next staging-next merge.

**Comment by abbradar** [comment]
Well, I usually avoid crowding staging with not-mass-rebuildy PRs to ease bisect in case it's needed (searching for problems among mass rebuild commits can be painful). 

On July 17, 2019 2:23:15 PM GMT+03:00, Timo Kaufmann <notifications@github.com> wrote:
>Why not just retarget this PR to staging and merge now? Then we'll get
>it automatically on the next staging-next merge.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/64716#issuecomment-512213020

-- 
Nikolay.

**Comment by timokau** [comment]
Fair enough.

**Comment by timokau** [comment]
I tried to build tensorflow with SSE4.1 support, but unfortunately the build failed. Apparently that has to be specified at configure time now. This is my attempt to get it working:
https://github.com/timokau/nixpkgs/commits/3323284
The patch to buildBazelPackages is necessary since tensorflow won't accept the `--config=opt` flag for fetching.

Feel free to cherry-pick and/or rebase.

**Comment by timokau** [comment]
@abbradar in case you haven't noticed yet, the tensorflow build (py2 and py3, but there are different issues for py2 and py3 I think) is [currently](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.python37Packages.tensorflow.x86_64-linux) failing. Unfortunately I don't have access to the necessary compute resources to debug this right now.

**Comment by abbradar** [comment]
Fixed in 71095460b94b6c382afe188d1caa0b7048030b8a. Thanks for the heads up!

**Pull request merged**
